File: TP01/q17/PalindromoR.c
#include <stdio.h>
#include <string.h>

// função recursiva para verificar se uma string é um palíndromo
int verificarPalindromoRecursivo(char texto[], int inicio, int fim) {
    //ignora caracteres esopeciais 
    while (texto[inicio] > '~') 
        inicio++;
    while (texto[fim] > '~') 
        fim--;

    if (inicio >= fim) {
        return 1; //caso base: se os índices se cruzarem ou forem iguais, é um palíndromo
    }

    if (texto[inicio] != texto[fim]) {
        return 0; //se forem diferentes retorna falso
    }

    return verificarPalindromoRecursivo(texto, inicio + 1, fim - 1); //chamada recursiva com índices ajustados
}

int main() {
    char texto[500];

    //"flag" de continuação para o loop
    int continuar = 1;
    while (continuar && fgets(texto, 500, stdin)) { //enquanto continuar e houver linhas de entrada
        int tamanho = strlen(texto);

        if (tamanho > 0 && texto[tamanho - 1] == '\n') { //se a ultima posição for quebra de linha, substitui por "null"
            texto[tamanho - 1] = '\0';
        }
        
        //se a string for "FIM" troca o estado de continuar para 0
        if (strcmp(texto, "FIM") == 0) {
            continuar = 0;
        } else {
            int ehPalindromo = verificarPalindromoRecursivo(texto, 0, strlen(texto) - 1);
            if (ehPalindromo) {
                printf("SIM\n");
            } else {
                printf("NAO\n");
            }
        }
    }

    return 0;
}

File: TP01/q16/PalindromoR.java
package q16;

import java.util.*;

public class PalindromoR {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in); //scanner para ler a entrada
        String texto;
        
        // continua lendo enquanto houver linhas e a entrada não for "FIM"
        while (scanner.hasNextLine() && !(texto = scanner.nextLine()).equals("FIM")) {
            boolean ehPalindromo = verificarPalindromo(texto, 0, texto.length() - 1);
            
            if (ehPalindromo) { //imprime a resposta
                System.out.println("SIM");
            } else {
                System.out.println("NAO");
            }
        }
        
        scanner.close();
    }
    
    public static boolean verificarPalindromo(String texto, int inicio, int fim) {
        if (inicio >= fim) { //caso base: se os índices se cruzarem ou forem iguais
            return true; //se chegou aqui, é um palíndromo
        }
        
        if (texto.charAt(inicio) != texto.charAt(fim)) { //compara os caracteres
            return false; //se forem diferentes retorna falso
        }
        
        return verificarPalindromo(texto, inicio + 1, fim - 1); //chamada recursiva com índices ajustados
    }
}

